zipkin-collector/src/test/scala/com/twitter/zipkin/collector/BlockingItemQueueTest.scala

```diff
@@ -21,23 +20,46 @@ class BlockingItemQueueTest extends FunSuite {
   }
 
   test("Sleeps on a max queue and waits for the worker to drain") {
-
+    val expectedItemCount = 12
+    val queueSize = 10
     val stats:InMemoryStatsReceiver = new InMemoryStatsReceiver()
-    val queue = new BlockingItemQueue[Unit, Unit](10, 1, fallingBehindWorker, 100.millis,
-                                                  100.millis, stats)
+    val queue = new BlockingItemQueue[Unit, Unit](queueSize, 1, fallingBehindWorker, 100.millis, 100.millis, stats)
+
+
     // Add 11 and not 10 because the first one that's going to be added will be consumed right away
-    assert(Await.result(fill(queue, 11)))
-    assert(Await.ready(queue.add(Item)).poll.get.isReturn)
+    assert(Await.result(fill(queue, queueSize + 1)))
+
+    // add an item, expect the queueFull stat to get incremented we need this to return because
+    // the blocking queue will block because its full, so we use a future
+    FuturePool.unboundedPool { queue.add(Item) }
+    // poll for when the queueFull counter gets incremented or timeout
+    poll( () => stats.counter("queueFull").apply() >= 1 )
+    // Items are sitting in the queue blocking and not processing, this unblocks them
+    latch.countDown()
+
+    // Wait for all items to be processed
+    poll( () => queue.size() == 0 )
     Await.ready(queue.close())
-    assert(stats.counter("queueFull").apply() >= 1)
-    assert(stats.counter("successes").apply() == 12)
+
+    // All items should be processed
+    poll( () => stats.counter("successes").apply() == expectedItemCount )
     assert(queue.size() == 0)
+    assert(stats.counter("queueFull").apply() >= 1)
+  }
+
+  def poll[A](f: () => Boolean): Boolean = {
+    val func = FuturePool.unboundedPool { do_poll(f) }
```

Comment:
yeah was thinking like `for (1 to count) if (foo.apply()) return true; else sleep 1000` in scala. The thread stuff is distracting.